src/ucp/core/ucp_mm.c

```diff
@@ -291,100 +273,200 @@ static inline int ucp_mem_map_is_allocate(const ucp_mem_map_params_t *params)
+        if (status != UCS_OK) {
+            return status;
+        }
 
-    ucs_debug("unmapping buffer %p memh %p", memh->address, memh);
+        reg_address = mem.address;
+        reg_length  = mem.length;
+    } else if (context->config.ext.reg_whole_alloc_bitmap & UCS_BIT(mem_type)) {
+        ucp_memory_detect_internal(context, address, length, &mem_info);
+        reg_address = mem_info.base_address;
+        reg_length  = mem_info.alloc_length;
+    } else {
+        reg_address = address;
+        reg_length  = length;
+    }
 
-    /* Unregister from all memory domains */
-    alloc_md_memh = UCT_MEM_HANDLE_NULL;
-    status = ucp_mem_rereg_mds(context, 0, NULL, 0, 0, memh->alloc_md, memh->mem_type,
-                               &alloc_md_memh, memh->uct, &memh->md_map);
-    if (status != UCS_OK) {
-        goto out;
+    if (context->rcache != NULL) {
+        status = ucs_rcache_get(context->rcache, reg_address, reg_length,
+                                PROT_READ|PROT_WRITE, NULL, &rregion);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        region = ucs_derived_of(rregion, ucp_mem_rcache_region_t);
+
+        reg_address = (void*)rregion->super.start;
+        reg_length  = rregion->super.end - rregion->super.start;
+    } else {
+        region = ucs_calloc(1, sizeof(ucp_mem_rcache_region_t) +
+                            sizeof(uct_mem_h) * context->num_mds, "ucp rcache");
+        if (!region) {
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        region->super.super.start   = (intptr_t)reg_address;
+        region->super.super.end     = (intptr_t)reg_address + reg_length;
+        region->memh.alloc_md_index = UCP_NULL_RESOURCE;
+        region->memh.alloc_method   = UCT_ALLOC_METHOD_LAST;
     }
 
-    /* If the memory was also allocated, release it */
-    if (memh->alloc_method != UCT_ALLOC_METHOD_LAST) {
-        mem.address = memh->address;
-        mem.length  = memh->length;
-        mem.method  = memh->alloc_method;
-        mem.md      = memh->alloc_md;  /* May be NULL if method is not MD */
-        mem.memh    = alloc_md_memh;   /* May be INVALID if method is not MD */
+    if (alloc) {
+        if (mem.method == UCT_ALLOC_METHOD_MD) {
+            for (md_index = 0; md_index < context->num_mds; md_index++) {
+                if (mem.md == context->tl_mds[md_index].md) {
+                    region->memh.alloc_md_index = md_index;
+                    region->memh.uct[md_index]  = mem.memh;
+                    region->memh.md_map        |= UCS_BIT(md_index);
+                    break;
+                }
+            }
+        }
 
-        status = uct_mem_free(&mem);
-        if (status != UCS_OK) {
-            goto out;
+        region->memh.alloc_method = mem.method;
+        region->memh.mem_type     = mem.mem_type;
+    } else {
+        region->memh.mem_type     = mem_type;
+    }
+
+    new_md_map = ~region->memh.md_map & reg_md_map;
```

Comment:
if (likely(memh map contains md map)) return ...
otherwise - slow path function to register all MDs